HotChocolate source generator is broken in Visual Studio 2022

### Product

Hot Chocolate

### Version

15.1.0-p.12

### Link to minimal reproduction

https://github.com/sunghwan2789/hotchocolate/tree/repro/type-analyzers

### Steps to reproduce

1. `git clone https://github.com/sunghwan2789/hotchocolate.git --branch repro/type-analyzers --single-branch hotchocolate-repro`
2. `.\hotchocolate-repro\hotchocolate.slnx`
3. build solution in Visual Studio 2022

### What is expected?

Analyzer generates `IRequestExecutorBuilder AddServerTypes(this IRequestExecutorBuilder builder)` extension method.

### What is actually happening?

Visual Studio fails to load `HotChocolate.Types.Analyzers` and do nothing.

### Relevant log output

```shell
Build started at 오전 10:51...
1>------ Build started: Project: Server, Configuration: Debug Any CPU ------
1>Skipping analyzers to speed up the build. You can execute 'Build' or 'Rebuild' command to run analyzers.
1>CSC : warning CS8032: An instance of analyzer HotChocolate.Types.Analyzers.GraphQLServerGenerator cannot be created from D:\.nuget\packages\hotchocolate.types.analyzers\15.1.0-p.12\analyzers\dotnet\cs\HotChocolate.Types.Analyzers.dll : Could not load file or assembly 'System.Runtime, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. 지정된 파일을 찾을 수 없습니다..
1>D:\source\repos\hotchocolate\Server\Program.cs(5,5,5,19): error CS1061: 'IRequestExecutorBuilder' does not contain a definition for 'AddServerTypes' and no accessible extension method 'AddServerTypes' accepting a first argument of type 'IRequestExecutorBuilder' could be found (are you missing a using directive or an assembly reference?)
1>Done building project "Server.csproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
========== Build completed at 오전 10:51 and took 00.552 seconds ==========
```

### Additional context

_No response_

--- a/src/HotChocolate/Core/src/Execution/DependencyInjection/SchemaRequestExecutorBuilderExtensions.Relay.cs
+++ b/src/HotChocolate/Core/src/Execution/DependencyInjection/SchemaRequestExecutorBuilderExtensions.Relay.cs
@@ -19,7 +19,11 @@ public static partial class SchemaRequestExecutorBuilderExtensions
     public static IRequestExecutorBuilder AddGlobalObjectIdentification(
         this IRequestExecutorBuilder builder)
     {
-        builder.AddDefaultNodeIdSerializer();
+        if (builder.Services.All(t => t.ServiceType != typeof(INodeIdSerializer)))
+        {
+            builder.AddDefaultNodeIdSerializer();
+        }
+
         return builder.ConfigureSchema(c => c.AddGlobalObjectIdentification());
     }
 
@@ -39,7 +43,11 @@ public static partial class SchemaRequestExecutorBuilderExtensions
         this IRequestExecutorBuilder builder,
         bool registerNodeInterface)
     {
-        builder.AddDefaultNodeIdSerializer();
+        if (builder.Services.All(t => t.ServiceType != typeof(INodeIdSerializer)))
+        {
+            builder.AddDefaultNodeIdSerializer();
+        }
+
         return builder.ConfigureSchema(c => c.AddGlobalObjectIdentification(registerNodeInterface));
     }
 

--- a/src/HotChocolate/Core/test/Types.Tests/Types/Relay/NodeIdSerializerTests.cs
+++ b/src/HotChocolate/Core/test/Types.Tests/Types/Relay/NodeIdSerializerTests.cs
@@ -1,0 +1,42 @@
+using Microsoft.Extensions.DependencyInjection;
+
+namespace HotChocolate.Types.Relay;
+
+public class NodeIdSerializerTests
+{
+    [Fact]
+    public void Registered_NodeIdSerializer_Should_Not_Be_Overwritten_By_AddGlobalObjectIdentification()
+    {
+        // arrange
+        var provider = new ServiceCollection()
+            .AddGraphQL()
+            .AddLegacyNodeIdSerializer()
+            .AddGlobalObjectIdentification()
+            .Services.BuildServiceProvider();
+
+        // act
+        var serializer = provider.GetRequiredService<INodeIdSerializer>();
+
+        // assert
+        Assert.IsType<LegacyNodeIdSerializer>(serializer);
+    }
+
+    [Fact]
+    public void
+        Registered_DefaultNodeIdSerializer_With_OutputNewIdFormat_Should_Not_Be_Overwritten_By_AddGlobalObjectIdentification()
+    {
+        // arrange
+        var provider = new ServiceCollection()
+            .AddGraphQL()
+            .AddDefaultNodeIdSerializer(outputNewIdFormat: false)
+            .AddGlobalObjectIdentification()
+            .Services.BuildServiceProvider();
+
+        // act
+        var serializer = provider.GetRequiredService<INodeIdSerializer>();
+        var serializedId = serializer.Format("Foo", 32);
+
+        // assert
+        Assert.Equal("Rm9vCmkzMg==", serializedId);
+    }
+}

